### PR TITLE
Experimental webhook token authentication support

### PR DESCRIPTION
**Pull request by olesku**
This PR adds experimental support for webhook token authentication, https://kubernetes.io/docs/admin/authentication/#webhook-token-authentication

To use add the following config:

```yaml
experimental:
 authentication:
   webhook:
     enabled: true
     cacheTTL: 1m0s
     configBase64: base64-encoded-webhook-yaml
```

configBase64 is a base64 encoded webhook-yaml configuration.

### CODE REVIEW

File: config/templates/cloud-config-controller [hunk 1]
```diff
@@ -795,13 +799,19 @@ write_files:
           - mountPath: /etc/ssl/certs
             name: ssl-certs-host
             readOnly: true
+          - mountPath: /etc/kubernetes/webhooks
```
Comment (gianrubio) [inline config/templates/cloud-config-controller]: Just mount the volume if the webHook is enabled

File: config/templates/cloud-config-controller [hunk 2]
```diff
@@ -406,7 +406,7 @@ write_files:
       kind: ConfigMap
       apiVersion: v1
       metadata:
-        name: calico-config 
+        name: calico-config
```
Comment (gianrubio) [inline config/templates/cloud-config-controller]: If you won't need to change this lines, put back to the old format

File: config/templates/cloud-config-controller [hunk 3]
```diff
@@ -564,7 +564,7 @@ write_files:
       ---
 
       apiVersion: extensions/v1beta1
-      kind: Deployment 
+      kind: Deployment
```
Comment (gianrubio) [inline config/templates/cloud-config-controller]: Same as above

File: config/templates/cloud-config-controller [hunk 4]
```diff
@@ -795,13 +799,19 @@ write_files:
...
+            name: kubernetes-webhooks
+            readOnly: true
         volumes:
         - hostPath:
             path: /etc/kubernetes/ssl
           name: ssl-certs-kubernetes
         - hostPath:
             path: /usr/share/ca-certificates
           name: ssl-certs-host
+        - hostPath:
```
Comment (gianrubio) [inline config/templates/cloud-config-controller]: Same as above

File: config/templates/cluster.yaml [hunk 5]
```diff
@@ -475,6 +475,11 @@ kmsKeyArn: "{{.KMSKeyARN}}"
 #     # See https://github.com/coreos/kube-aws/issues/230 for more information.
 #     rbac:
 #       enabled: true
+#     authentication:
```
Comment (gianrubio) [inline config/templates/cluster.yaml]: Write few lines commenting how to enable this auth or maybe just a link is enough

File: config/templates/cloud-config-controller (hunk 2 above)
Comment (olesku) [inline config/templates/cloud-config-controller]: Im not sure I agree with you on this, what good does those whitespaces do ?

File: config/templates/cluster.yaml (hunk 5 above)
Comment (olesku) [inline config/templates/cluster.yaml]: I agree, i've added a link to Kubernetes' documentation about webhook token authentication

File: config/templates/cloud-config-controller (hunk 1 above)
Comment (olesku) [inline config/templates/cloud-config-controller]: Good catch. Fixed now

**Comment by mumoshu** [comment]
Hi @olesku, thanks for the PR!
Just curious but how could we test this feature?
Is there a reference implementation of a webhook authentication endpoint/service in the wild for testing purpose?
Other than that, LGTM overall 👍

**Comment by olesku** [comment]
Hi @mumoshu, thanks for your feedback!
For my particular usecase i've implemented a custom webhook service that communicates with the IDP that my company use. Not sure if there is a reference implementation out in the wild, tried to google for it, but wasn't able to find one.

However there are multiple webhook projects for integrating with different authentication services, one example (for ldap) is: https://github.com/apprenda-kismatic/kubernetes-ldap

**Comment by mumoshu** [comment]
Thanks @olesku!
The code LGTM.
I'm going to merge this to see how it works for everyone 👍

**Comment by mumoshu** [comment]
Merged. Thanks for your contribution @olesku

**Pull request merged**